cmd/osm-controller/osm-controller.go

```diff
@@ -210,23 +210,48 @@ func main() {
 	// TODO(draychev): figure out the NS and POD
 	metricsStore := metricsstore.NewMetricStore("TBD_NameSpace", "TBD_PodName")
 
-	// Expose /debug endpoints and data only if the enableDebugServer flag is enabled
-	var debugServer debugger.DebugServer
-	if cfg.IsDebugServerEnabled() {
-		debugServer = debugger.NewDebugServer(certDebugger, xdsServer, meshCatalog, kubeConfig, kubeClient, cfg, kubernetesClient)
-	}
-
 	funcProbes := []health.Probes{xdsServer}
 	httpProbes := getHTTPHealthProbes()
-	httpServer := httpserver.NewHTTPServer(funcProbes, httpProbes, metricsStore, constants.MetricsServerPort, debugServer)
+	httpServer := httpserver.NewHTTPServer(funcProbes, httpProbes, metricsStore, constants.MetricsServerPort)
 	httpServer.Start()
 
-	// Wait for exit handler signal
-	<-stop
+	// Expose /debug endpoints and data only if the enableDebugServer flag is enabled
+	debugServerRunning := !cfg.IsDebugServerEnabled()
+	debugImpl := debugger.NewDebugImpl(certDebugger, xdsServer, meshCatalog, kubeConfig, kubeClient, cfg, kubernetesClient)
+	debugServer := httpserver.NewDebugHTTPServer(debugImpl, constants.DebugPort)
+
+	errCh := make(chan error)
+	go configureDebugServer(debugServer, debugImpl, debugServerRunning, cfg, errCh)
+
+	select {
+	case err := <-errCh:
+		log.Error().Err(err).Msg("Unable to configure debug server")
+
+		// Wait for exit handler signal
+	case <-stop:
+	}
 
 	log.Info().Msg("Goodbye!")
 }
 
+func configureDebugServer(debugServer *httpserver.DebugServer, debugImpl debugger.DebugServer, debugServerRunning bool, cfg configurator.Configurator, errCh chan error) {
+	for range cfg.GetAnnouncementsChannel() {
+		if debugServerRunning && !cfg.IsDebugServerEnabled() {
+			debugServerRunning = false
+			err := debugServer.Stop()
+			if err != nil {
+				log.Error().Err(err).Msg("Unable to stop debug server")
+				errCh <- err
+			}
+		}
+		if !debugServerRunning && cfg.IsDebugServerEnabled() {
+			debugServerRunning = true
+			debugServer = httpserver.NewDebugHTTPServer(debugImpl, constants.DebugPort)
+			debugServer.Start()
+		}
+	}
+}
```

Comment:
I admit I don't know much about how the announcements channel works, but if there's a scenario where someone is rapid-fire toggling the debug server and the queue is filled with start,stop,start,stop,start,stop is there a way to process whatever the latest state is without having to do all the intermediate starting and stopping of the server?

Reply:
The rate of announcements are controlled by the ticker in `repeater.go`. The bigger problem I see with the current implementation is that it will break the moment there are more than 1 subscribers listening for announcement on a single channel.

Reply:
@shashankbarsin - yea we need to fix that but I'd prefer if we thought through how we want to handle this kind of thing generically in all places where there is an announcement channel and then handled it after we come up with a strategy.